### ISSUE TITLE
Bizhawk throws an exeption when movie folder contains alot of movies

### ISSUE THREAD

**Issue by GoogleCodeExporter** [TASVideos/BizHawk#435]
```
What steps will reproduce the problem?
1. Have a selection of movies in the movies folder
2. 10 is enough, but the more you have the easier to trigger the bug
3.

What is the expected output? What do you see instead?
When loading a movie, the expected result is to see the relevant movie file for 
me to play and/or continue.

This along standing bug that's annoyed me for ages, but it time to get it known 
before a new bizhawk release is put out.

Sometimes, more often than not. If I as I build up a list of movie files, 
bizhawk will through up an exception after it has tried to scan through all the 
movie files.

Normally the exception window can be closed and just hit the rescan button and 
the relevant movie will eventually show.  Sometimes it can take 2 or 3 re 
attempts but will get their in the end.

What version of the product are you using? On what operating system?

Any recent bizhawk version, Windows 8.1 64bit

Please provide any additional information below.

Its almost as if bizhawk is unzipping every movie file to search for the 
correct hash but it takes too long and bizhawk gets fed up of waiting and 
causes an exception.


```

Original issue reported on code.google.com by `MThompso...@gmail.com` on 16 Jun 2015 at 8:09

Attachments:
* [bizcrash.png](https://storage.googleapis.com/google-code-attachments/bizhawk/issue-435/comment-0/bizcrash.png)
* [bizcrash.txt](https://storage.googleapis.com/google-code-attachments/bizhawk/issue-435/comment-0/bizcrash.txt)

**Comment by zeromus** [comment]
https://github.com/TASVideos/BizHawk/commit/0341d24bc894f930fc2033ec710587fd25730c42 should have fixed this, anxiously awaiting confirmation

**Comment by hegyak** [comment]
This error appears sometimes:

************** Exception Text **************
System.AggregateException: One or more errors occurred. ---> System.IO.DirectoryNotFoundException: Could not find a part of the path 'D:\Emu\BizHawk\Movies\TAStudio states'.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.Directory.DeleteHelper(String fullPath, String userPath, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound)
   at System.IO.Directory.Delete(String fullPath, String userPath, Boolean recursive, Boolean checkHost)
   at System.IO.Directory.Delete(String path, Boolean recursive)
   at BizHawk.Client.Common.TasStateManager..ctor(TasMovie movie)
   at BizHawk.Client.Common.TasMovie..ctor(String path, Boolean startsFromSavestate, BackgroundWorker progressReportWorker)
   at BizHawk.Client.Common.MovieService.Get(String path)
   at BizHawk.Client.EmuHawk.PlayMovie.PreLoadMovieFile(HawkFile hf, Boolean force)
   at BizHawk.Client.EmuHawk.PlayMovie.AddMovieToList(String filename, Boolean force)
   at BizHawk.Client.EmuHawk.PlayMovie.<>c__DisplayClass7.<ScanFiles>b__5(Int32 i)
   at System.Threading.Tasks.Parallel.<>c__DisplayClassf`1.<ForWorker>b__c()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask)
   at System.Threading.Tasks.Task.<>c__DisplayClass11.<ExecuteSelfReplicating>b__10(Object param0)
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.For(Int32 fromInclusive, Int32 toExclusive, Action`1 body)
   at BizHawk.Client.EmuHawk.PlayMovie.ScanFiles()
   at BizHawk.Client.EmuHawk.PlayMovie.PlayMovie_Load(Object sender, EventArgs e)
   at System.Windows.Forms.Form.OnLoad(EventArgs e)
   at System.Windows.Forms.Form.OnCreateControl()
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl()
   at System.Windows.Forms.Control.WmShowWindow(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
   at System.Windows.Forms.ContainerControl.WndProc(Message& m)
   at System.Windows.Forms.Form.WmShowWindow(Message& m)
   at System.Windows.Forms.Form.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.OnMessage(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
---> (Inner Exception #0) System.IO.DirectoryNotFoundException: Could not find a part of the path 'D:\Emu\BizHawk\Movies\TAStudio states'.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.Directory.DeleteHelper(String fullPath, String userPath, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound)
   at System.IO.Directory.Delete(String fullPath, String userPath, Boolean recursive, Boolean checkHost)
   at System.IO.Directory.Delete(String path, Boolean recursive)
   at BizHawk.Client.Common.TasStateManager..ctor(TasMovie movie)
   at BizHawk.Client.Common.TasMovie..ctor(String path, Boolean startsFromSavestate, BackgroundWorker progressReportWorker)
   at BizHawk.Client.Common.MovieService.Get(String path)
   at BizHawk.Client.EmuHawk.PlayMovie.PreLoadMovieFile(HawkFile hf, Boolean force)
   at BizHawk.Client.EmuHawk.PlayMovie.AddMovieToList(String filename, Boolean force)
   at BizHawk.Client.EmuHawk.PlayMovie.<>c__DisplayClass7.<ScanFiles>b__5(Int32 i)
   at System.Threading.Tasks.Parallel.<>c__DisplayClassf`1.<ForWorker>b__c()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask)
   at System.Threading.Tasks.Task.<>c__DisplayClass11.<ExecuteSelfReplicating>b__10(Object param0)<---

**Comment by zeromus** [comment]
Loading a TasMovie causes a TasStateManager to get built. TasStateManager sets up temp folders on the user's disk. In other words, it's not a read-only operation. The multi-threaded movie scanning blithely assumes that it's a read-only operation and there should be no threading issues.

In this case TasStateManager always mounts the same temp folder. Consequently, no more than one TasStateManager can get created at a time. This is a threading issue.

Since no more than one at a time can exist, the creation of that object would need to be tightly managed and not created internally by TasMovies. 

But it would be better to change TasStateManager to use a random temp directory. I wouldn't base it on the GUID of the movie, either. It should be truly standalone per TasStateManager.

Moreover, I don't think a TasMovie should be creating a TasStateManager unless state cache related operations are happening. If this was the case then, regardless of how many temp directories there are, loading a TasMovie becomes a read-only operation once more.  Having TasMovie be powerful enough to start meddling with the user's disk when it's constructed spooks me.

This is adelikat's bug, but I'll fix it after I give him some time to think about it.

**Comment by zeromus** [comment]
i think this was fixed by 3ec314d and b981e00

**Comment by zeromus** [comment]
This has been addressed in at least half a dozen different ways. Closing to wait for a new report.
